src/features/ImeiProgramming/index.tsx

```diff
@@ -0,0 +1,415 @@
+                <span
+                    className={`mdi tw-leading-none ${
+                        !show ? ' mdi-eye-off' : ' mdi-eye'
+                    }`}
+                />
+            </button>
+        </div>
+    );
+};
+
+const isSupportedDevice = (deviceInfo?: DeviceInfo) =>
+    deviceInfo?.jlink?.protectionStatus === 'NRFDL_PROTECTION_STATUS_NONE' &&
+    deviceInfo?.jlink?.deviceVersion?.toLocaleUpperCase().match(/NRF91\d1/);
+
+const getStatus = async (device?: Device, deviceInfo?: DeviceInfo) => {
+    if (!device || !deviceInfo) return 'Not connected';
+    if (deviceInfo.jlink?.protectionStatus !== 'NRFDL_PROTECTION_STATUS_NONE')
+        return 'PROTECTED';
+
+    if (!device.serialNumber || !isSupportedDevice(deviceInfo))
+        return 'UNSUPPORTED';
+
+    const defaultImeiNumber = 'FFFFFFFFFFFFFFF';
+    try {
+        const res = await readIMEI(device.serialNumber);
+        if (res !== defaultImeiNumber) {
+            return 'IMEI_AVAILABLE';
+            return 'IMEI_SET';
+        }
+        return 'IMEI_AVAILABLE';
+    } catch (e) {
+        return 'FIRMWARE';
+    }
+};
+
+const updateDeviceInfo = async (device: Device) => {
+    const deviceInfo = await NrfutilDeviceLib.deviceInfo(device);
+    setSelectedDeviceInfo(deviceInfo);
+    return deviceInfo;
+};
+
+const recover = async (device: Device) => {
+    try {
+        await NrfutilDeviceLib.batch()
+            .recover('Application')
+            .reset('Application')
+            .run(device);
+        return getStatus(device, await updateDeviceInfo(device));
+    } catch (e) {
+        // TODO
+    }
+    return 'TODO';
+};
+
+const programFirmware = async (device: Device) => {
+    await NrfutilDeviceLib.program(
+        device,
+        getAppFile(path.join('resources', 'firmware', 'pcm.zip')),
+        () => {},
+        'Modem',
+        { reset: 'RESET_SYSTEM' }
+    );
+    const newDeviceInfo = await updateDeviceInfo(device);
+    return getStatus(device, newDeviceInfo);
+};
+
+export default () => {
+    const device = useSelector(selectedDevice);
+    const deviceInfo = useSelector(selectedDeviceInfo);
+    const [status, setStatus] = useState('NONE');
+    const [showSpinner, setShowSpinner] = useState(false);
+    const [useCloud, setUseCloud] = useState(false);
+    const [apiKey, setAPIKey] = useState('');
+    const [manualIMEI, setManualIMEI] = useState('');
+    const [cloudIMEI, setCloudIMEI] = useState('');
+    const [hasProgrammedFirmware, setHasProgrammedFirmware] = useState(false);
+    const [error, setError] = useState<string>();
+
+    const waitForAction = async (action: () => Promise<void>) => {
+        setShowSpinner(true);
+        setError(undefined);
+        await action().finally(() => {
+            setShowSpinner(false);
+        });
+    };
+
+    useEffect(() => {
+        // if device disconnects
+        if (!device && status !== 'NONE') {
+            setStatus('NONE');
+        }
+    }, [device, status]);
```

Comment:
I get your point but the "IMEI copy because consumed" notice will be visible when relevant. If the user didn't copy it and disconnects the device then idk.